sys/net/gnrc/network_layer/ipv6/ext/frag/gnrc_ipv6_ext_frag.c

```diff
@@ -46,17 +60,256 @@ void gnrc_ipv6_ext_frag_init(void)
+            remaining -= ptr->size;
+        }
+        if (last) {
+            last->next = ptr;
+        }
+        last = ptr;
+        ptr = ptr->next;
+    }
+    assert(nh != NULL);
+    /* then the fragment header */
+    ptr = gnrc_ipv6_ext_build(last, last->next, *nh, sizeof(ipv6_ext_frag_t));
+    if (ptr == NULL) {
+        DEBUG("ipv6_ext_frag: unable to create fragmentation header\n");
+        gnrc_pktbuf_release(to_send);
+        _snd_buf_free(snd_buf);
+        return;
+    }
+    remaining -= sizeof(ipv6_ext_frag_t);
+    frag_hdr = ptr->data;
+    ipv6_ext_frag_set_offset(frag_hdr, snd_buf->offset);
+    if (!last_fragment) {
+        ipv6_ext_frag_set_more(frag_hdr);
+    }
+    frag_hdr->id = byteorder_htonl(snd_buf->id);
+    *nh = PROTNUM_IPV6_EXT_FRAG;
+    last = ptr;
+    /* then the rest */
+    while (remaining && snd_buf->pkt) {
+        if (last_fragment ||
+            (snd_buf->pkt->size <= remaining)) {
+            ptr = snd_buf->pkt;
+            snd_buf->pkt = ptr->next;
+            ptr->next = NULL;
+        }
+        else {
+            ptr = gnrc_pktbuf_mark(snd_buf->pkt, remaining,
+                                   GNRC_NETTYPE_UNDEF);
+            if (ptr == NULL) {
+                DEBUG("ipv6_ext_frag: packet buffer full, canceling fragmentation\n");
+                gnrc_pktbuf_release(to_send);
+                _snd_buf_free(snd_buf);
+                return;
+            }
+            assert(snd_buf->pkt->next == ptr);  /* we just created it with mark */
+            snd_buf->pkt->next = snd_buf->pkt->next->next;
+            ptr->next = NULL;
+        }
+        last->next = ptr;
+        last = ptr;
+        remaining -= ptr->size;
+        snd_buf->offset += ptr->size;
+    }
+    assert(len != NULL);
+    /* adapt IPv6 header length field */
+    *len = byteorder_htons(gnrc_pkt_len(to_send->next->next));
+    msg.type = GNRC_IPV6_EXT_FRAG_SEND_FRAG;
+    msg.content.ptr = to_send;
+    msg_try_send(&msg, gnrc_ipv6_pid);
+    if (last_fragment) {
+        _snd_buf_del(snd_buf);
+    }
+    else {
+        msg.type = GNRC_IPV6_EXT_FRAG_SEND;
+        msg.content.ptr = snd_buf;
+        msg_try_send(&msg, gnrc_ipv6_pid);
+    }
+}
+
+static gnrc_ipv6_ext_frag_send_t *_snd_buf_alloc(void)
+{
+    for (unsigned i = 0; i < GNRC_IPV6_EXT_FRAG_SEND_SIZE; i++) {
+        gnrc_ipv6_ext_frag_send_t *snd_buf = &_snd_bufs[i];
+        if (snd_buf->pkt == NULL) {
+            return snd_buf;
+        }
+    }
+    return NULL;
+}
+
+static void _snd_buf_del(gnrc_ipv6_ext_frag_send_t *snd_buf)
+{
+    snd_buf->per_frag = NULL;
+    snd_buf->pkt = NULL;
+}
+
+static void _snd_buf_free(gnrc_ipv6_ext_frag_send_t *snd_buf)
+{
+    if (snd_buf->per_frag) {
+        gnrc_pktbuf_release(snd_buf->per_frag);
+    }
+    if (snd_buf->pkt) {
+        gnrc_pktbuf_release(snd_buf->pkt);
+    }
+    _snd_buf_del(snd_buf);
+}
+
+static gnrc_pktsnip_t *_determine_last_per_frag(gnrc_pktsnip_t *ptr)
```

Comment:
Just a small description of what this function does would be nice.

Reply:
Added documentation to its definition above.